micropython/lora/lora/sx1276.py

```diff
@@ -0,0 +1,592 @@
+
+        return old_value
+
+    def _dio_isr(self, pin):
+        if pin not in self._irq_en:
+            # TODO: This check against _irq_en shouldn't really be necessary,
+            # but weirdly DIO1 (configured for RxTimeout) seems to go high
+            # during TX, for no apparent reason (it's also masked out in
+            # _IRQ_FLAGS_MASK). Pull-up/down on the pin doesn't change it. This
+            # is probably a driver bug, and once it's fixed then _irq_en can be
+            # fully removed as the modem IRQ mask should determine which IRQs
+            # are driven high
+            if _DEBUG:
+                print("ISR ignored {}".format(pin))
+            return
+        if _DEBUG:
+            print("ISR {}".format(pin))
+        if self._irq_callback:
+            self._irq_callback()
+
+    def start_receive(self, timeout_ms=None):
+        mode = self._get_mode()
+
+        self._set_mode(_OPMODE_MODE_STDBY)
+
+        self._reg_write(_REG_IRQ_FLAGS_MASK, 0xFF & ~(_IRQ_RX_DONE | _IRQ_RX_TIMEOUT))
+
+        if self._dio0:
+            self._reg_update(
+                _REG_DIO_MAPPING1, _DIO0_MAPPING_MASK << _DIO0_MAPPING_SHIFT, 0
+            )  # DIO0 = RXDone
+
+        if self._dio1:
+            self._reg_update(
+                _REG_DIO_MAPPING1, _DIO1_MAPPING_MASK << _DIO1_MAPPING_SHIFT, 0
+            )  # DIO1 = RXTimeout
+        self._irq_en = (self._dio0, self._dio1)
+
+        if timeout_ms is not None:
+            timeout_syms = (timeout_ms + self._t_sym_ms - 1) // self._t_sym_ms
+            timeout_syms = _clamp(timeout_syms, 4, 1023)
+            self._reg_update(
+                _REG_MODEM_CONFIG2,
+                _MODEM_CONFIG2_SYMB_TIMEOUT_MSB_MASK,
+                timeout_syms >> 8,
+            )
+            self._reg_write(_REG_SYMB_TIMEOUT_LSB, timeout_syms & 0xFF)
+
+        # Allocate the full FIFO for RX
+        self._reg_write(_REG_FIFO_ADDR_PTR, 0)
+        self._reg_write(_REG_FIFO_RX_BASE_ADDR, 0)
+        self._set_mode(
+            _OPMODE_MODE_RX_CONTINUOUS if timeout_ms is None else _OPMODE_MODE_RX_SINGLE
+        )
+
+        # will_irq if DIO0 and DIO1 both hooked up, or DIO0 and no timeout
+        return self._dio0 and (self._dio1 or timeout_ms is None)
+
+    def is_receive_completed(self):
+        # assumption here that _REG_IRQ_FLAGS_MASK is set correctly, so we don't have to check exactly which flags are set
+        return self._reg_read(_REG_IRQ_FLAGS)
+
+    def end_receive(self, timeout_ms):
+        self._irq_en = ()
+        if timeout_ms is None:
+            # We need to use continuous mode for a no-timeout receive, so stop receiving now
+            self._set_mode(_OPMODE_MODE_STDBY)
+
+        flags = self._reg_update(
+            _REG_IRQ_FLAGS,
+            _IRQ_RX_DONE | _IRQ_RX_TIMEOUT,
+            _IRQ_RX_DONE | _IRQ_RX_TIMEOUT,
+        )  # clear flags
+        if flags & _IRQ_RX_TIMEOUT:  # Modem timed out
+            return None
```

Comment:
This code assumes that end_receive is being called because receive-complete got signalled. What if the upper layer just wants to TX a packet and thus calls end_receive to stop the current RX mode? As far as I can tell, it's incorrect to just call start_transmit...

Reply:
There is now a "low level API" with documentation and (I hope) less arbitrary behaviour.